robot/basestation/static/js/roslibjs/ros-helpers.js

```diff
@@ -188,6 +204,99 @@ function initRosWeb () {
+  let secsFloored = Math.floor(secs) // seconds after truncating
+  let nsecs = Math.round(1000000000*(secs-secsFloored)) // nanoseconds since the previous second
+
+  // return a dictionary for the ROS log and a float for the gui console
+  return [{secs : secsFloored, nsecs : nsecs}, currentTime]
+}
+
+function logGeneric (logLevel, message) {
+  stamps = getTimestamp()
+  rosTimestamp = stamps[0] // dictionary for the ROS log message
+  consoleTimestamp = stamps[1].toString().split(' ')[4] // hh:mm:ss from date object
+
+  // @TODO cedric can help me refactor this
+  switch (logLevel) {
+    case ROSDEBUG:
+      consoleMsg = "[DEBUG] " + '[' + consoleTimestamp + ']' + ': ' + message
+      console.log(consoleMsg)
+      appendToConsole(consoleMsg, false)
+      break
+    case ROSINFO:
+      consoleMsg = "[INFO] " + '[' + consoleTimestamp + ']' + ': ' + message
+      console.log(consoleMsg)
+      appendToConsole(consoleMsg, false)
+      break
+    case ROSWARN:
+      consoleMsg = "[WARN] " + '[' + consoleTimestamp + ']' + ': ' + message
+      console.error(consoleMsg)
+      appendToConsole(consoleMsg, false)
+      break
+    case ROSERROR:
+      consoleMsg = "[ERROR] " + '[' + consoleTimestamp + ']' + ': ' + message
+      console.error(consoleMsg)
+      appendToConsole(consoleMsg, false)
+      break
+    case ROSFATAL:
+      consoleMsg = "[FATAL] " + '[' + consoleTimestamp + ']' + ': ' + message
+      console.error(consoleMsg)
+      appendToConsole(consoleMsg, false)
+      break
+  }
+
+  // @TODO bonus if we can determine how to generate a log file for the gui
+  // rather than simply having it show up in rosout.log
+  // an attempt was made with the name and file elements in the log message
+  ros_logger.publish(
+    new ROSLIB.Message({
+      // I'm only publishing the essentials. We could include more info if so desired
+      header : {
+        // seq // uint32: sequence ID, seems to increment automatically
+        stamp : rosTimestamp
+        // frame_id // string: probably only useful for tf
+      },
+      level : logLevel,
+      // name : '/web_gui', // name of the node
+      msg : message,
+      // file : 'web_gui' // string: we could specify the js file that generated the log
+      // function // string: we could specify the parent function that generated the log
+      // line // uint32: we could specify the specific line of code that generated the log
+      // topics // string[]: topic names that the node publishes
+    })
+  )
+}
+
+// Copying how rospy handles the logging functions:
+// All log messages should go to the log file (but how?)
+// debug messages only go to /rosout if a value is set while initializing the js
+// info goes to stdout, warn-error-fatal go to stderr
+// rospy also takes in *args and **kwargs, how does this work and do I need to implement that?
+function logDebug (message) {
+  // @TODO only log debug messages if we set a parameter somewhere.
```

Comment:
See CONTRIBUTING.md

NO TODOS :angry: :angry:

Reply:
i will create a list of all the todos/bonus things and we can discuss with peter how we wish to procede

Reply:
removed all the todos from the code and put them in this PR description so you can think about it with @PeterGhimself